src/lib/y2users/users_simple/reader.rb

```diff
@@ -38,8 +38,19 @@ def read_to(config)
           # lets just use the strongest available
           user.password = Password.new(config, user["uid"],
             value: Yast::Builtins.cryptsha512(user["userPassword"]))
+          user.password.plain_value = user["userPassword"]
           config.users << user
         end
+
+        # Read also root user settings
+        root_pwd_plain = Yast::UsersSimple.GetRootPassword
+        root_pwd = Password.new(config, "root",
+          value: Yast::Builtins.cryptsha512(user["userPassword"]))
```

Comment:
If, at this point, we only know the plain password I would only set `plain_value`. In my opinion, the transformation of that into a final `value` is not responsibility of this reader.

Reply:
This introduces a risk of getting both passwords out of sync if something like this is executed (creative usage of the API at some points for clarity):

```ruby
initial = Y2Users::Config.new(:initial)
# The following call should only set #plain_password, otherwise...
Y2Users::UsersSimple::Reader.new.read_to(initial)

config2 = initial.clone
# Now we introduce inconsistencies between both attributes
config2.root_user.password.plain_value = "S3cr3t"

# This should be the point in which #value should be calculated based on #plain_password
# it may even use a different crypt algorithm
Y2Users::Whatever::Writer.new(config2).write
```

Reply:
In that regard, I would even consider that maybe the `Password#plain_value=` setter should set `#value` to nil.

Reply:
Thinking about the problem of both values going out of sync... maybe we should have just one value and a separate attribute to indicate whether such value is plain or already encrypted. Something like this (just a first wild idea):

```ruby
class Password
  attr_reader :value

  def initialize(value, plain: true)
    @value = value
    @plain_value = plain
  end

  def plain?
    !!@plain_value
   end

  def plain_value
    return nil unless plain?
    value
  end

  def encrypted_value
    return nil if plain?
    value
   end
end
```

Reply:
It makes sense for me to have just one. At first I consider if we can do it without flag, but it is a bit problematic as value is even more tricky as it can be even something like disable login or even more tricky value like "!<value>" which is locked password with old value after ! which can be used to unlock account. So I am not sure how to show this domain difficulty into API. As we have 

- plain value
- encrypted value
- locked encrypted value
- disabled login ( by now represented in that encrypted value and sadly it is  "!" or "*" which can be valid plain passwords. )

I am even not sure in your snippet if we should keep that value reader which is context dependent.
Maybe we should instead use something like Enum in value? And that object answer its kind from kinds above and provide value when it makes sense.

So what about something like
```
class PasswordValue
  secret_attribute :value

  def initialize(value)
      @value = value
  end

  def encrypted?
     false
  end

  def plain?
      false
  end

  def locked?
      false
  end

  def disabled?
      false
  end
end

class EncryptedPassword < PasswordValue
   def initialize(value)
     super
   end

  def encrypted?
      true
  end
end

class PlainPassword
   def initialize(value)
     super
   end

  def plain?
      true
  end
end

...
```

Reply:
As I understand this, the idea is to then also have subclasses for `LockedPassword` and `DisabledPassword`. If that's the case, I see some overlapping between the `PasswordValue#value` and the class hierarchy. For example, what if I do this?

```ruby
example = EncryptedPassword.new("*")
```

Moreover, if I want to create a `DisabledPassword` object, which `value` should I pass to the constructor?

Or maybe I understood it wrong and the idea is to only have `EncryptedPassword` and `PlainPassword`, with an implementation of `EncryptedPassword` similar to this:

```ruby
class EncryptedPassword < PasswordValue
   def initialize(value)
     super
   end

  def encrypted?
      true
  end

  def locked?
    value && value.start_with("!")
  end

  def disabled?
    value == "*"
  end
end
```

If that's the case, I like it more. But I'm not sure if having the `Y2Users::Password` and then a `Y2Users::PasswordValue` may be a bit over the top.

Reply:
> As I understand this, the idea is to then also have subclasses for `LockedPassword` and `DisabledPassword`. If that's the case, I see some overlapping between the `PasswordValue#value` and the class hierarchy. For example, what if I do this?
> 
> ```ruby
> example = EncryptedPassword.new("*")
> ```

this should raise as it is invalid encrypted value

> 
> Moreover, if I want to create a `DisabledPassword` object, which `value` should I pass to the constructor?

nothing, it should have empty constructor and pass that "*" inside it like
```ruby
def initialize
  super("*")
end
```

> 
> Or maybe I understood it wrong and the idea is to only have `EncryptedPassword` and `PlainPassword`, with an implementation of `EncryptedPassword` similar to this:
> 
> ```ruby
> class EncryptedPassword < PasswordValue
>    def initialize(value)
>      super
>    end
> 
>   def encrypted?
>       true
>   end
> 
>   def locked?
>     value && value.start_with("!")
>   end
> 
>   def disabled?
>     value == "*"
>   end
> end
> ```
> 
> If that's the case, I like it more. But I'm not sure if having the `Y2Users::Password` and then a `Y2Users::PasswordValue` may be a bit over the top.

Yeah, I do not like naming...But I am not sure if we should rename Y2Users::Password to Y2Users::Shadow and let live just PasswordValue? BTW if it helps, we can also have that subclasses as nested ones and have helper constructors in PasswordValue like
```ruby
class PasswordValue
   def self.create_encrypted(value)
      PasswordValue::EncryptedPassword.new(value)
   end
end
```

Reply:
IMHO, something like what @ancorgs proposes should be enough, shouldn't it?

~~~
class Password
  secret_attribute :value

  def initialize(value, plain: true)
    self.value = value
    @plain = plain
  end

  def plain?
    !!@plain
  end

  def encrypted?
    !plain?
  end

  def locked?
    value.start_with("!")
  end

  def disabled?
    value == "*"
  end
end
~~~

And maybe, I would not allow to modify a password. We can create a new one if we need a new password. Otherwise, we could still end with wrongly configured passwords:

~~~
pwd = Password.new("S3cr3t")
pwd.value = encrypt("My new password")
pwd.plain? #=> true
~~~

Reply:
What I do not like about @ancorgs proposal is that it is hard to keep it consistent as you see with setting value. In fact even your example has bug as if value is "!" it is disabled and not locked. Locked is only if it start with "!" but follows with old encrypted code. Also it lets a lot of responsibility to Password configuration, but it is all related password value type. So instead of trying to make value and all flags consistent, own object is always consistent as it prevent to have mixture of different value types.

Reply:
BTW if I take that 
```
pwd.change("!1afads34fes", :encrypted)
pwd.encypted? #=> true
pwd.locked? #=> true
```
and instead use object it is more readable for me:
```
pwd.value = Password.encrypted_value("!1afads34fes")
pwd.value.encypted? #=> true
pwd.value.locked? #=> true
```